# 5. Métricas de Avaliação e Comparação

## Introdução

Nesta etapa serão utilizadas métricas adicionais para avaliar os modelos de regressão e classificação desenvolvidos anteriormente.

O objetivo é analisar o desempenho dos modelos de forma mais detalhada, utilizando diferentes métricas além das apresentadas na Parte 3.

Para regressão serão utilizadas:

- MAE (Mean Absolute Error)
- RMSE (Root Mean Squared Error)
- R²

Para classificação serão utilizadas:

- Accuracy
- Precision
- Recall
- F1-Score

In [1]:
# importando as bibliotecas

import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeClassifier

from sklearn.metrics import (
    r2_score,
    accuracy_score,
    mean_absolute_error,
    mean_squared_error,
    precision_score,
    recall_score,
    f1_score
)

pd.set_option('display.max_columns', None)
pd.options.display.float_format = '{:.2f}'.format

In [2]:
# carregando os dados

df = pd.read_csv('train.csv')

print(df.shape)
df.head()

(1460, 81)


,Id,MSSubClass,MSZoning,LotFrontage,LotArea,Street,Alley,LotShape,LandContour,Utilities,LotConfig,LandSlope,Neighborhood,Condition1,Condition2,BldgType,HouseStyle,OverallQual,OverallCond,YearBuilt,YearRemodAdd,RoofStyle,RoofMatl,Exterior1st,Exterior2nd,MasVnrType,MasVnrArea,ExterQual,ExterCond,Foundation,BsmtQual,BsmtCond,BsmtExposure,BsmtFinType1,BsmtFinSF1,BsmtFinType2,BsmtFinSF2,BsmtUnfSF,TotalBsmtSF,Heating,HeatingQC,CentralAir,Electrical,1stFlrSF,2ndFlrSF,LowQualFinSF,GrLivArea,BsmtFullBath,BsmtHalfBath,FullBath,HalfBath,BedroomAbvGr,KitchenAbvGr,KitchenQual,TotRmsAbvGrd,Functional,Fireplaces,FireplaceQu,GarageType,GarageYrBlt,GarageFinish,GarageCars,GarageArea,GarageQual,GarageCond,PavedDrive,WoodDeckSF,OpenPorchSF,EnclosedPorch,3SsnPorch,ScreenPorch,PoolArea,PoolQC,Fence,MiscFeature,MiscVal,MoSold,YrSold,SaleType,SaleCondition,SalePrice
0,1,60,RL,65.00,8450,Pave,NaN,Reg,Lvl,AllPub,Inside,Gtl,CollgCr,Norm,Norm,1Fam,2Story,7,5,2003,2003,Gable,CompShg,VinylSd,VinylSd,BrkFace,196.00,Gd,TA,PConc,Gd,TA,No,GLQ,706,Unf,0,150,856,GasA,Ex,Y,SBrkr,856,854,0,1710,1,0,2,1,3,1,Gd,8,Typ,0,NaN,Attchd,2003.00,RFn,2,548,TA,TA,Y,0,61,0,0,0,0,NaN,NaN,NaN,0,2,2008,WD,Normal,208500
1,2,20,RL,80.00,9600,Pave,NaN,Reg,Lvl,AllPub,FR2,Gtl,Veenker,Feedr,Norm,1Fam,1Story,6,8,1976,1976,Gable,CompShg,MetalSd,MetalSd,NaN,0.00,TA,TA,CBlock,Gd,TA,Gd,ALQ,978,Unf,0,284,1262,GasA,Ex,Y,SBrkr,1262,0,0,1262,0,1,2,0,3,1,TA,6,Typ,1,TA,Attchd,1976.00,RFn,2,460,TA,TA,Y,298,0,0,0,0,0,NaN,NaN,NaN,0,5,2007,WD,Normal,181500
2,3,60,RL,68.00,11250,Pave,NaN,IR1,Lvl,AllPub,Inside,Gtl,CollgCr,Norm,Norm,1Fam,2Story,7,5,2001,2002,Gable,CompShg,VinylSd,VinylSd,BrkFace,162.00,Gd,TA,PConc,Gd,TA,Mn,GLQ,486,Unf,0,434,920,GasA,Ex,Y,SBrkr,920,866,0,1786,1,0,2,1,3,1,Gd,6,Typ,1,TA,Attchd,2001.00,RFn,2,608,TA,TA,Y,0,42,0,0,0,0,NaN,NaN,NaN,0,9,2008,WD,Normal,223500
3,4,70,RL,60.00,9550,Pave,NaN,IR1,Lvl,AllPub,Corner,Gtl,Crawfor,Norm,Norm,1Fam,2Story,7,5,1915,1970,Gable,CompShg,Wd Sdng,Wd Shng,NaN,0.00,TA,TA,BrkTil,TA,Gd,No,ALQ,216,Unf,0,540,756,GasA,Gd,Y,SBrkr,961,756,0,1717,1,0,1,0,3,1,Gd,7,Typ,1,Gd,Detchd,1998.00,Unf,3,642,TA,TA,Y,0,35,272,0,0,0,NaN,NaN,NaN,0,2,2006,WD,Abnorml,140000
4,5,60,RL,84.00,14260,Pave,NaN,IR1,Lvl,AllPub,FR2,Gtl,NoRidge,Norm,Norm,1Fam,2Story,8,5,2000,2000,Gable,CompShg,VinylSd,VinylSd,BrkFace,350.00,Gd,TA,PConc,Gd,TA,Av,GLQ,655,Unf,0,490,1145,GasA,Ex,Y,SBrkr,1145,1053,0,2198,1,0,2,1,4,1,Gd,9,Typ,1,TA,Attchd,2000.00,RFn,3,836,TA,TA,Y,192,84,0,0,0,0,NaN,NaN,NaN,0,12,2008,WD,Normal,250000


## Preparação dos Dados

Para manter a mesma lógica utilizada na Parte 3, serão utilizadas apenas as colunas numéricas do conjunto de dados, removendo registros com valores ausentes.

In [3]:
# selecionando apenas variaveis numericas

df_simples = df.select_dtypes(
    include=['int64', 'float64']
).dropna()

X = df_simples.drop(
    ['Id', 'SalePrice'],
    axis=1
)

y_regressao = df_simples['SalePrice']

## Avaliação do Modelo de Regressão Linear

O modelo de regressão linear será treinado novamente para que possamos calcular métricas adicionais de desempenho.

In [4]:
# divisao treino e teste

X_train_R, X_test_R, y_train_R, y_test_R = train_test_split(
    X,
    y_regressao,
    test_size=0.2,
    random_state=42
)

# treinamento

modelo_regressao = LinearRegression()

modelo_regressao.fit(
    X_train_R,
    y_train_R
)

# previsoes

previsoes_R = modelo_regressao.predict(
    X_test_R
)

# metricas

r2 = r2_score(
    y_test_R,
    previsoes_R
)

mae = mean_absolute_error(
    y_test_R,
    previsoes_R
)

rmse = np.sqrt(
    mean_squared_error(
        y_test_R,
        previsoes_R
    )
)

print('metricas da regressao linear')
print(f'mae: {mae:,.2f}')
print(f'rmse: {rmse:,.2f}')
print(f'r²: {r2:.4f}')

metricas da regressao linear
mae: 23,615.83
rmse: 39,852.60
r²: 0.7986


### Interpretação dos Resultados

O modelo de Regressão Linear apresentou um erro absoluto médio (MAE) de aproximadamente US$ 23.615,83, indicando que as previsões diferem dos valores reais em cerca de vinte e três mil dólares, em média.

O RMSE obtido foi de aproximadamente US$ 39.852,60, mostrando que erros maiores possuem um impacto mais significativo na avaliação do modelo.

O coeficiente de determinação (R²) foi de 0,7986, indicando que cerca de 79,86% da variação dos preços das casas pode ser explicada pelas variáveis utilizadas.

Os resultados demonstram que o modelo possui boa capacidade de previsão para o conjunto de dados analisado, sendo capaz de estimar os preços dos imóveis com um nível satisfatório de precisão.

## Avaliação do Modelo de Classificação

Agora será avaliado o modelo de Árvore de Decisão utilizado para classificar as casas em categorias de preço alto ou baixo.

In [5]:
# criando variavel binaria

mediana = df_simples['SalePrice'].median()

y_classificacao = (
    df_simples['SalePrice']
    > mediana
).astype(int)

# divisao treino e teste

X_train_C, X_test_C, y_train_C, y_test_C = train_test_split(
    X,
    y_classificacao,
    test_size=0.2,
    random_state=42
)

# treinamento

modelo_arvore = DecisionTreeClassifier(
    random_state=42
)

modelo_arvore.fit(
    X_train_C,
    y_train_C
)

# previsoes

previsoes_C = modelo_arvore.predict(
    X_test_C
)

# metricas

acuracia = accuracy_score(
    y_test_C,
    previsoes_C
)

precisao = precision_score(
    y_test_C,
    previsoes_C
)

recall = recall_score(
    y_test_C,
    previsoes_C
)

f1 = f1_score(
    y_test_C,
    previsoes_C
)

print('metricas da classificacao')
print(f'acuracia: {acuracia:.4f}')
print(f'precisao: {precisao:.4f}')
print(f'recall: {recall:.4f}')
print(f'f1-score: {f1:.4f}')

metricas da classificacao
acuracia: 0.8844
precisao: 0.8943
recall: 0.8943
f1-score: 0.8943


### Interpretação dos Resultados

O modelo de Árvore de Decisão apresentou acurácia de aproximadamente 88,44%, indicando que a maioria das casas foi classificada corretamente como preço alto ou preço baixo.

A precisão de 89,43% demonstra que, quando o modelo prevê uma casa como pertencente à categoria de preço alto, essa previsão tende a estar correta na maior parte das vezes.

O recall de 89,43% mostra que o modelo conseguiu identificar corretamente grande parte das casas pertencentes à categoria positiva.

Já o F1-Score de 89,43% confirma um bom equilíbrio entre precisão e recall, evidenciando a consistência do modelo de classificação.

Os resultados obtidos demonstram que a Árvore de Decisão apresentou excelente desempenho na tarefa de classificação dos imóveis.

## Comparação dos Modelos

Os dois modelos apresentaram resultados satisfatórios para os objetivos propostos.

A Regressão Linear alcançou um coeficiente de determinação (R²) de aproximadamente 0,80, demonstrando boa capacidade para prever o preço de venda dos imóveis com base em suas características.

Já a Árvore de Decisão apresentou acurácia de 88,44%, além de precisão, recall e F1-Score superiores a 89%, indicando excelente desempenho na classificação dos imóveis em categorias de preço alto e preço baixo.

Comparando os resultados, observa-se que ambos os modelos foram eficientes em suas respectivas tarefas. A Regressão Linear mostrou-se adequada para estimar valores de venda, enquanto a Árvore de Decisão apresentou elevado desempenho na identificação de imóveis de maior ou menor valor.

Dessa forma, os modelos atenderam aos objetivos propostos pelo projeto e contribuíram para uma melhor compreensão do comportamento do mercado imobiliário presente no conjunto de dados.